Add new API GET methods

blanked opened the pull request:
I thought it'll be good to have `GET` methods as well. This will allow admins to write automation (eg. reports for publishing user access info) on the plugin without having to resort to script console.

blanked commented:
ah it looks like the PR-merge build failed due to an issue on the agent? i'll close and reopen the PR to induce a rebuild

AbhyudayaSharma commented:
Thanks a lot @blanked ! I will try to review this PR as soon as possible.

kalyantag commented:
@AbhyudayaSharma , Any update on this feature?

AbhyudayaSharma commented:
@kalyantag Sorry, I got busy with other things and forgot about this PR. I will review it this weekend.

kalyantag commented:
Thanks @AbhyudayaSharma

AbhyudayaSharma commented on src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:
```diff
@@ -443,4 +446,131 @@ public void doRemoveSidFromAgentRole(@QueryParameter(required = true) String rol
+    @Restricted(NoExternalUse.class)
+    public void doGetAgentRole(@QueryParameter(required = true) String name) throws IOException {
+        Jenkins.get().checkPermission(Jenkins.ADMINISTER);
+        JSONObject responseJson = new JSONObject();
```
I think we can pass a `true` to the JSONObject constructor so that we return null instead of an empty object (`{}`). Or maybe we can return a descriptive error message saying that role does not exist.

http://json-lib.sourceforge.net/apidocs/jdk15/net/sf/json/JSONObject.html#JSONObject()

blanked replied:
Hmm..I would prefer not to return an error message and instead return either an empty object or null as I find it more inline with other services. I can change it to return null

AbhyudayaSharma commented on src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:
```diff
@@ -443,4 +446,131 @@ public void doRemoveSidFromAgentRole(@QueryParameter(required = true) String rol
+
+    /**
+     * API Method to get an {@link FolderRole}
+     * Example: {@code curl -X GET 'http://localhost:8080/jenkins/folder-auth/getFolderRole?name=folderRole1}
```
```suggestion
     * Example: {@code curl -X GET 'http://localhost:8080/jenkins/folder-auth/getFolderRole?name=folderRole1'}
```

AbhyudayaSharma commented on src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyAPI.java:
```diff
@@ -302,4 +305,69 @@ public static void removeSidFromAgentRole(String sid, String roleName) {
+        try {
+            strategy = (FolderBasedAuthorizationStrategy) jenkins.getAuthorizationStrategy();
+        } catch (ClassCastException error) {
+            throw new IllegalStateException("FolderBasedAuthorizationStrategy is not the" + " current authorization strategy");
```
```suggestion
            throw new IllegalStateException(Messages.FolderBasedAuthorizationStrategy_NotCurrentStrategy());
```

AbhyudayaSharma commented on src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:
```diff
@@ -443,4 +446,131 @@ public void doRemoveSidFromAgentRole(@QueryParameter(required = true) String rol
+        JSONObject responseJson = new JSONObject();
+        AgentRole role = FolderAuthorizationStrategyAPI.getAgentRole(name);
+        if (role != null) {
+            responseJson.put("name", role.getName());
```
Would it be possible to let json-lib handle converting the AbstractRole objects to JSONObjects using [`JSONObject#fromObject`](http://json-lib.sourceforge.net/apidocs/jdk15/net/sf/json/JSONObject.html#fromObject(java.lang.Object))?

blanked replied:
To use `JSONObject#fromObject`, the role class must be of type string (json formatted), map, etc. The current class doesn't so it'll require a refactor for the class to extend either of these classes it to work. I'll keep it as is? Or we can

AbhyudayaSharma commented on src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:
```diff
@@ -443,4 +446,131 @@ public void doRemoveSidFromAgentRole(@QueryParameter(required = true) String rol
+            responseJson.put("sids", role.getSids());
+            responseJson.put("permissions", getPermissionsFromRole(role));
+        }
+        Stapler.getCurrentResponse().setContentType("application/json;charset=UTF-8");
```
I think it is possible to directly return a `JSONObject` from a `doFoo` function like done here: https://github.com/jenkinsci/folder-auth-plugin/blob/5fc71be89c0560e09f6d03c3817c0738b8fb3fff/src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java#L274

blanked replied:
Oh I wasn't aware of that. Is that due to the `@GET` annotation from stapler?

AbhyudayaSharma replied:
`@GET` is to restrict the call to only GET requests: https://github.com/stapler/stapler/blob/master/core/src/main/java/org/kohsuke/stapler/verb/GET.java

I think returning a JSON response is implemented here: https://github.com/stapler/stapler/blob/0a23a9000de1dafb6a16fa883504697443cc05e3/core/src/main/java/org/kohsuke/stapler/json/JsonHttpResponse.java

AbhyudayaSharma commented on src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:
```diff
@@ -443,4 +446,131 @@ public void doRemoveSidFromAgentRole(@QueryParameter(required = true) String rol
+
+    /**
+     * API Method to get an {@link GlobalRole}
+     * Example: {@code curl -X GET 'http://localhost:8080/jenkins/folder-auth/getGlobalRole?name=admin}
```
```suggestion
     * Example: {@code curl -X GET 'http://localhost:8080/jenkins/folder-auth/getGlobalRole?name=admin'}
```

AbhyudayaSharma commented on src/test/java/io/jenkins/plugins/folderauth/FolderAuthorizationWebAPITest.java:
```diff
@@ -202,13 +206,145 @@ public void testDeleteAgentRole() throws IOException {
+
+        // Verifying that the method returns all 3 roles
+        String response = page.getWebResponse().getContentAsString();
+        JSONObject json = JSONObject.fromObject(response);
```
Is it possible to convert the `JSONObject` back to a `FolderRole` or `AbstractRole` object using http://json-lib.sourceforge.net/apidocs/jdk15/net/sf/json/JSONObject.html#toBean(net.sf.json.JSONObject,%20java.lang.Class) ?

blanked replied:
Similar to my comment before, the roles need to extend map/bean to work. It'll require a refactor of the existing class

AbhyudayaSharma commented on src/main/webapp/js/addrole.js:
```diff
@@ -21,7 +21,7 @@ const addGlobalRole = () => {
     }
 
-    sendPostRequest(`${rootURL}/folder-auth/addGlobalRole`, response);
+    sendPostRequest(rootURL + "/folder-auth/addGlobalRole", response);
```
Is there a reason for converting the template strings to concatenation using `+`?

blanked replied:
The template strings are unfortunately unsupported in the `WebClient` in JenkinsRule, which throws errors during testing. I've decided to remove template strings instead of using an alternative test suite.

AbhyudayaSharma reviewed (commented):
Hi @blanked! Thanks a lot for the PR. Sorry for the late review; I completely forgot about it. 

The PR looks good. I just have a few questions/suggestions. Please let me know what you think about them.

AbhyudayaSharma commented:
@blanked Thanks a lot! Looks like the build is failing because of missing `@throws` in JavaDocs.

```
Exit code: 1 - /home/jenkins/workspace/Plugins_folder-auth-plugin_PR-69/src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:459: warning: no @throws for java.io.IOException
```

blanked commented:
> 
> 
> @blanked Thanks a lot! Looks like the build is failing because of missing `@throws` in JavaDocs.
> 
> ```
> Exit code: 1 - /home/jenkins/workspace/Plugins_folder-auth-plugin_PR-69/src/main/java/io/jenkins/plugins/folderauth/FolderAuthorizationStrategyManagementLink.java:459: warning: no @throws for java.io.IOException
> ```

Oh the `throws IOException` was meant to be removed due to the removal of stapler writer but I missed it.

blanked commented:
Ok I've implemented the suggestions. Can you take another look and see if the PR is good to go? Thanks @AbhyudayaSharma

blanked commented:
Ok is there any way of retriggering the tests? It seems like there were problems on the agent for the build. Closing and reopening doesn't seem to trigger a new build

AbhyudayaSharma commented:
I think the build can be restarted by pushing an empty commit (`git commit --allow-empty`) or closing it, waiting for 2 minutes and then reopening. I also believe that there was a hook added to rebuild from comments on pull requests in last year's GSoC. I will try to find whether it is available on ci.jenkins.io

________________________________
From: Hui Jun <notifications@github.com>
Sent: Saturday, 6 March 2021, 21:17
To: jenkinsci/folder-auth-plugin
Cc: Abhyudaya Sharma; Mention
Subject: Re: [jenkinsci/folder-auth-plugin] Add new API GET methods (#69)


Ok is there any way of retriggering the tests? It seems like there were problems on the agent for the build. Closing and reopening doesn't seem to trigger a new build

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/jenkinsci/folder-auth-plugin/pull/69#issuecomment-791975089>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACXQV3YDJYIBITKSPEIJW4TTCJFB7ANCNFSM4S5SCAPA>.

blanked commented:
Ok I finally got a build to run successfully. I believe I've addressed all of the comments/suggestion. Let me know if there are any other comments/suggestions? Thanks!

AbhyudayaSharma commented:
Hi @oleg-nenashev ! If you have some time, could you please take a quick look at this PR?

AbhyudayaSharma commented:
Hi @blanked. Sorry for the delay. What would you think about exporting the entire plugin configuration as JSON as done in  https://github.com/jenkinsci/folder-auth-plugin/pull/58/files#diff-237f6cdc333b75a3ea591296a855f117edf78bb4ec1f9cbbc2391811e672304d?
